R/calc_coherence.R

```diff
@@ -0,0 +1,259 @@
+  #1 reduce tcm to top n word space
+  #2 create pairs of tcm reference indices of wi / wj to be used for coherence calculation
+  #3 get values for all wiwj pairs from tcm and calculate score for each pair given a calculation rule specfied by particular coherence measure
+  #     e.g. pmi = function(wi, wj, ndocs_tcm, tcm)  {log2((tcm[wi,wj]/ndocs_tcm) + 1e-12) - log2(tcm[wi,wi]/ndocs_tcm) - log2(tcm[wj,wj]/ndocs_tcm)}
+  #4 aggregate the results via mean over number of wiwj pairs (and if desired aggreagte over all topics)
+
+  #TODO
+  #(ii) use word vectors for wi / wj instead of single words, hence, subsets such as S_one_any, etc.
+  #     creating, e.g., one any subsets requires to store one index against a list of indices, hence, formulas need
+  #     adaption, e.g., something like tcm[unlist(wi), unlist(wj)] might work
+  #(iii)Currently indices of subsets are stored in memory, this might be turned to dynamic creation of indices to spare memory usage
+  #     the lines topic_coherence[,tcm_idxs := split(match... would have to be incorporated into create_wiwj_set
+
+#CREDITS / REFERENCES -------------------------------------------------
+  #The following paper is the main theoretical basis for this code
+  #https://dl.acm.org/citation.cfm?id=2685324
+  #Authors: Roeder, Michael; Both, Andreas; Hinneburg, Alexander (2015)
+  #Title: Exploring the Space of Topic Coherence Measures.
+  #In: Xueqi Cheng, Hang Li, Evgeniy Gabrilovich und Jie Tang (Hg.):
+  #Proceedings of the Eighth ACM International Conference on Web Search and Data Mining - WSDM '15.
+  #the Eighth ACM International Conference. Shanghai, China, 02.02.2015 - 06.02.2015.
+  #New York, New York, USA: ACM Press, S. 399-408.
+  #The paper has already been implemented as a Java implementation "palmetto" that exceeds the current functionality of this R implementation
+  #Main Author / Maintainer: Michael Roeder
+  #https://github.com/dice-group/Palmetto
+  #http://aksw.org/Projects/Palmetto.html
+
+  #some ideas for getting/ordering/accessing indices of tcm have been inspired by packages stm and textmineR
+  #(although, these packages use different functions to achieve the same result)
+  #stm: Molly Roberts, Brandon Stewart and Dustin Tingley
+  #https://github.com/bstewart/stm/blob/master/R/semanticCoherence.R
+  #textmineR: Tommy Jones
+  #https://github.com/TommyJones/textmineR/blob/master/R/CalcProbCoherence.R
+
+#INITIAL----------------------------------------------------------------------------
+  top <- as.vector(top_term_matrix)
+  top_unq <- unique(top)
+  ntopics <- ncol(top_term_matrix)
+  nterms <- nrow(top_term_matrix)
+  topic_coherence <- data.table(Topic = paste0("T", 1:ntopics))
+
+#PREPARE TCM------------------------------------------------------------------------
+  #check if input tcm really contains the same elements in rows and columns
+  if ( !all.equal(rownames(tcm)[order(rownames(tcm))], colnames(tcm)[order(colnames(tcm))]) ) {
```

Comment:
Nice, I learnt something very useful, have never used `stopifnot(setqual...))`, thanks. 

I added `ncol(tcm) == nrow(tcm)` as second condition, since we would not like duplicate entries which `setequal` would not detect. 

Done via 627e17e